overlord/assertstate/validation_set_tracking.go

```diff
@@ -0,0 +1,130 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package assertstate
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/snapcore/snapd/overlord/state"
+)
+
+// ValidationSetMode reflects the mode of respective validation set, which is
+// either monitoring or enforcing.
+type ValidationSetMode int
+
+const (
+	Monitor ValidationSetMode = iota
+	Enforce
+)
+
+// ValidationSetTracking holds tracking parameters for associated validation set.
+type ValidationSetTracking struct {
+	AccountID string            `json:"account"`
+	Name      string            `json:"name"`
+	Mode      ValidationSetMode `json:"mode"`
+
+	// PinnedAt is an optional pinned sequence point, or 0 if not pinned.
+	PinnedAt int `json:"pinned-at,omitempty"`
+
+	// Current is the current sequence point.
+	Current int `json:"cuurent,omitempty"`
+}
+
+// ValidationSetKey formats the given account id and name into a validation set key.
+func ValidationSetKey(accountID, name string) string {
+	return fmt.Sprintf("%s/%s", accountID, name)
+}
+
+// UpdateValidationSet updates ValidationSetTracking.
+// The method assumes valid tr fields.
+func UpdateValidationSet(st *state.State, tr *ValidationSetTracking) {
+	var vsmap map[string]*json.RawMessage
+	err := st.Get("validation-set-tracking", &vsmap)
+	if err != nil && err != state.ErrNoState {
+		panic("internal error: cannot unmarshal validation-set-tracking state: " + err.Error())
+	}
+	if vsmap == nil {
+		vsmap = make(map[string]*json.RawMessage)
+	}
+	data, err := json.Marshal(tr)
+	if err != nil {
+		panic("internal error: cannot marshal validation set tracking data: " + err.Error())
+	}
+	raw := json.RawMessage(data)
+	key := ValidationSetKey(tr.AccountID, tr.Name)
+	vsmap[key] = &raw
+	st.Set("validation-set-tracking", vsmap)
+}
+
+// DeleteValidationSet deletes a validation set for the given accoundID and name.
+// It is not an error to delete a non-existing one.
+func DeleteValidationSet(st *state.State, accountID, name string) error {
+	var vsmap map[string]*json.RawMessage
+	err := st.Get("validation-set-tracking", &vsmap)
+	if err != nil && err != state.ErrNoState {
+		panic("internal error: cannot unmarshal validation-set-tracking state: " + err.Error())
+	}
+	if len(vsmap) == 0 {
+		return nil
+	}
+	delete(vsmap, ValidationSetKey(accountID, name))
+	st.Set("validation-set-tracking", vsmap)
+	return nil
+}
+
+// GetValidationSet retrieves the ValidationSetTracking for the given account and name.
+func GetValidationSet(st *state.State, accountID, name string, tr *ValidationSetTracking) error {
+	if tr == nil {
+		return fmt.Errorf("internal error: tr is nil")
+	}
+
+	*tr = ValidationSetTracking{}
+
+	var vset map[string]*json.RawMessage
+	err := st.Get("validation-set-tracking", &vset)
+	if err != nil {
+		return err
+	}
+	key := ValidationSetKey(accountID, name)
+	raw, ok := vset[key]
+	if !ok {
+		return state.ErrNoState
+	}
+	err = json.Unmarshal([]byte(*raw), &tr)
```

Comment:
Here `&tr` is a pointer to a pointer to a `ValidationSetTracking`, is this desired? I was under the impression that we want to set what the caller has provided to us, not our local pointer to it. Note that due to how the API looks like, if `Unmarshal` really re-directs the pointer to something else that is not propagated to the caller.

Reply:
This is modeled 1:1 after snapstate.Get()

Reply:
I know and I used the same "trick" but I fear we don't exactly know why. This is cargo-culted from snapstate for a reason, that I think, is related to a bug in older JSON marshaling implementations. Today the code passes both with and without the ampersand here, indicating it is not strictly required in all cases. I think we should add a comment both here and in snapstate, explaining whey exactly this is being done.